californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java

```diff
@@ -192,195 +211,262 @@ public void onAcknowledgement() {
 			if (status.isComplete()) {
 				// clean up blockwise status
-				LOGGER.log(Level.FINE, "Ongoing finished on first block {0}", status);
-				exchange.setResponseBlockStatus(null);
-				exchange.setBlockCleanupHandle(null);
+				LOGGER.log(Level.FINE, "peer has requested last block of blockwise transfer: {0}", status);
+				clearBlock2Status(key);
 			} else {
-				LOGGER.log(Level.FINE, "Ongoing started {0}", status);
+				LOGGER.log(Level.FINE, "peer has requested intermediary block of blockwise transfer: {0}", status);
 			}
 
 			exchange.setCurrentResponse(block);
 			lower().sendResponse(exchange, block);
+		}
+	}
 
-		} else {
+	/**
+	 * Invoked when a response is sent to a peer.
+	 * <p>
+	 * This method initiates a blockwise transfer if the response's payload
+	 * exceeds {@code MAX_MESSAGE_SIZE}.
+	 * 
+	 * @param exchange The exchange the response is part of.
+	 * @param response The response to send to the peer.
+	 */
+	@Override
+	public void sendResponse(final Exchange exchange, final Response response) {
+
+		Response responseToSend = response;
+
+		if (isTransparentBlockwiseHandlingEnabled()) {
+
+			BlockOption requestBlock2 = exchange.getRequest().getOptions().getBlock2();
+			BlockOption responseBlock2 = response.getOptions().getBlock2();
+
+			if (requestBlock2 != null && requestBlock2.getNum() > 0) {
+
+				// peer has issued a random block access request
+
+				if (responseBlock2 != null) {
+
+					// the resource implementation supports blockwise retrieval (indicated by the
+					// presence of the block2 option in the response)
+
+					if (requestBlock2.getNum() != responseBlock2.getNum()) {
+						LOGGER.log(
+							Level.WARNING,
+							"resource [{0}] implementation error, peer requested block {1} but resource returned block {2}",
+							new Object[]{exchange.getRequest().getURI(), requestBlock2.getNum(), responseBlock2.getNum()});
+						responseToSend = Response.createResponse(exchange.getRequest(), ResponseCode.INTERNAL_SERVER_ERROR);
+						responseToSend.setType(response.getType());
+						responseToSend.setMID(response.getMID());
+					}
+
+				} else if (response.hasBlock(requestBlock2)) {
+
+					// the resource implementation does not support blockwise retrieval
+					// but instead has responded with the full response body
+					// crop the response down to the requested block
+					Block2BlockwiseStatus.crop(responseToSend, requestBlock2);
+
+				} else {
+
+					// peer has requested a non existing block
+					responseToSend = Response.createResponse(exchange.getRequest(), ResponseCode.BAD_OPTION);
+					responseToSend.setType(response.getType());
+					responseToSend.setMID(response.getMID());
+					responseToSend.getOptions().setBlock2(requestBlock2);
+
+				}
+
+			} else if (requiresBlockwise(exchange, response, requestBlock2)) {
+
+				KeyUri key = getKey(exchange, response);
+				Block2BlockwiseStatus status = getOutboundBlock2Status(key, exchange, response);
```

Comment:
When we receive multiple requests for the same resource coming from the same client (the scenario described above) then we will need to assume that the client wants to *start over* with the blockwise download. Based on what I wrote above, the only thing that would need to be changed is that we do not use `status.getNextResponseBlock()` (with no arg) but instead use `status.getNextResponseBlock(BlockOption)` with a newly instantiated BlockOption as the arg (indicating that we want to get the *first* response block). Explicitly creating a new instance (replacing a potentially already existing one) would also work but requires more code, I guess.